crates/edr_evm/src/mempool.rs

```diff
@@ -25,7 +25,7 @@ where
     /// Removes all pending transactions of the account corresponding to the
     /// provided address.
     pub fn remove_caller(&mut self, caller: &Address) -> Option<Vec<OrderedTransaction>> {
-        self.transactions.remove(caller)
+        self.transactions.shift_remove(caller)
```

Comment:
Question: I'm wondering what's the reason to change to `shift_remove`? Based on skimming the docs `remove` is `O(1)` while `shift_remove` is `O(n)`. Can that be a problem? Are we potentially calling this in a loop making it `O(n^2)`?

Reply:
I guess it's because the data structure, [according to its docs](https://crates.io/crates/indexmap):

> Preserves insertion order as long as you don't call .remove(), .swap_remove(), or other methods that explicitly change order. The alternate .shift_remove() does preserve relative order.

I wonder if we have a test for the scenario where:

1) Automining is disabled
2) Mempool is used in FIFO mode
3) 5 transactions are added to the mempool
4) 3rd transaction is removed
5) The rest of the txs preserve the order

Reply:
> I'm wondering what's the reason to change to shift_remove?

`remove` was deprecated because it was silently reordering. Instead they added two explicit methods that signal reordering vs order-preserving

Reply:
Thanks, I checked now and it seems that we're using `remove_caller` in a loop causing worst case `O(n^2)` removal:

https://github.com/NomicFoundation/hardhat/blob/69e340761ce52812fe720689bcec972345fc8df7/crates/edr_evm/src/miner.rs#L134-L138

We should turn that into worst case linear before merging this imo.

Reply:
> Thanks, I checked now and it seems that we're using `remove_caller` in a loop causing worst case `O(n^2)` removal:
> 
> https://github.com/NomicFoundation/hardhat/blob/69e340761ce52812fe720689bcec972345fc8df7/crates/edr_evm/src/miner.rs#L134-L138
> 
> We should turn that into worst case linear before merging this imo.

The order needs to be preserved, as Franco pointed out, so the entries have to be shifted.

Reply:
Technically we only need to preserve ordering when the mempool is in FIFO mode. But I think it's fine for this to be quadratic, N here is the number of transactions in the mempool, which is very unlikely to be a large number.

Reply:
Ok created tech debt issue based on in-person discussion: https://github.com/NomicFoundation/edr/issues/382